server/test/lib/scheduler/Scheduler.spec.ts

```diff
@@ -0,0 +1,294 @@
+      schedule.assignWorkspaceToUser(USER_ID_2, "2");
+
+      const promise = scheduler.filterByWhetherNotYetWorkedOn(workspaces);
+      promise.then(result => {
+        expect(result).to.have.deep.members(workspaces.filter(w => (
+          w.id !== "1-1" && w.id !== "2"
+        )));
+        done();
+      }).catch(e => {
+        done(e);
+      });
+    });
+  });
+
+  describe("filterByWhetherHasRemainingBudget", () => {
+    it("works in straightforward case", () => {
+      const result = scheduler.filterByWhetherHasRemainingBudget(workspaces);
+      expect(result).to.have.deep.members(workspaces.filter(w => (
+        w.id !== "5-4"
+      )));
+    });
+  });
+
+  describe("getIdsOfWorkspacesThatCouldBeNext", () => {
+    it("works in a straightforward case", done => {
+      schedule.assignWorkspaceToUser(USER_ID_1, "1-1");
+      fakeClock.tick(ONE_MINUTE / 2);
+      schedule.assignWorkspaceToUser(USER_ID_2, "2");
+
+      const promise = scheduler.getIdsOfWorkspacesThatCouldBeNext(USER_ID_1);
+      promise.then(result => {
+        expect(result).to.have.deep.members(["3", "4", "5", "5-1", "5-2", "5-3", "5-4"]);
+        done();
+      }).catch(e => {
+        done(e);
+      });
+    });
+
+    it("works when all workspaces assigned at least once", done => {
+      fakeClock.tick(100);
+      schedule.assignWorkspaceToUser(USER_ID, "1-1");
+      fakeClock.tick(100);
+      schedule.assignWorkspaceToUser(USER_ID, "2");
+      fakeClock.tick(100);
+      schedule.assignWorkspaceToUser(USER_ID, "2-1");
+      fakeClock.tick(100);
+      schedule.assignWorkspaceToUser(USER_ID, "5-1");
+      fakeClock.tick(100);
+      schedule.assignWorkspaceToUser(USER_ID, "1");
+      fakeClock.tick(100);
+      schedule.assignWorkspaceToUser(USER_ID, "4");
+      fakeClock.tick(100);
+      schedule.assignWorkspaceToUser(USER_ID, "3");
+      fakeClock.tick(100);
+      schedule.assignWorkspaceToUser(USER_ID, "2");
+
+      // NOTE 5-1 is excluded because it's currently being worked on!
+      const promise = scheduler.getIdsOfWorkspacesThatCouldBeNext(USER_ID_1);
+      promise.then(result => {
+        expect(result).to.have.deep.members(["5", "5-2", "5-3", "5-4"]);
+        done();
+      }).catch(e => {
+        done(e);
+      });
+    });
+  });
+
+  describe("findNextWorkspace", () => {
+    it("works in a straightforward case", done => {
+      scheduler.findNextWorkspace(USER_ID_1).then(() => {
+        const promise = scheduler.getIdOfCurrentWorkspace(USER_ID_1);
+        promise.then(result => {
+          expect(result).to.be.oneOf([
+            "1",
+            "1-1",
+            "1-1-1",
+            "2",
+            "2-1",
+            "2-2",
+            "3",
+            "4",
+            "5",
+            "5-1",
+            "5-2",
+            "5-3",
+            "5-4",
+          ]);
+          done();
+        }).catch(e => {
+          done(e);
+        });
+      });
+    });
+  });
+
+});
```

Comment:
You can definitely DRY up a bunch of these tests. No need to stay so close to native chai syntax.

Reply:
I think they've been DRYed up some since you posted this. Let me know if there are any specific additional improvements you have in mind.